### PR TITLE
Introduce Pytest, Tox, Coverage

### LINKED ISSUE #13: Testing Framework
I'd suggest that we start to use Pytest as a testing platform: http://pytest.org/ I've made _very_ good experiences with it.

If you agree, I could add the configuration for a nice test suite (including style checks).

Things like a mocked Minecraft server could be added as testing fixtures.

### PR DESCRIPTION
**Pull request by dbrgn**
Fixes #13.

**Still work in progress.** Qemu on Travis is very slow :/

The question is whether we actually need Qemu with a real server, or if we can go with mocking one. Then we could get rid of the Qemu setup.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env python
+import io
+from setuptools import setup, find_packages
+from minecraft import meta
+
+readme = io.open('README.rst', mode='r', encoding='utf8').read()
```
Comment (ghickman) [inline setup.py]: This can use `with open()`.

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env python
+import io
+from setuptools import setup, find_packages
+from minecraft import meta
```
Comment (ghickman) [inline setup.py]: Can you avoid importing the library please, importing the library at install time is a nasty anti-pattern. A plain text file which is read in will be fine for this project.

**Comment by ghickman** [comment]
@dbrgn – This is a good start but I don't think we need to target 3.4 as Raspbian doesn't ship it yet. Thanks for making the change to Tox!

**Comment by doismellburning** [comment]
That said, I think we still want to work with 3.4 (for when Raspbian _does_ ship it), but we _need_ it to work for 3.2 (Wheezy stable)

File: setup.py (hunk 2 above)
Comment (dbrgn) [inline setup.py]: As long as I just import the meta module, which in turn doesn't import anything from the library, it should be fine, right? The idea was that the code also has access to the program version (e.g. for debug output).

But I can also remove it entirely and just keep the version and other meta-data in the setup.py file :)

File: setup.py (hunk 1 above)
Comment (dbrgn) [inline setup.py]: The default `open` is not a good idea because it does not always handle encoding correctly, see https://github.com/strongswan/swidGenerator/pull/35. If the system has a terminal which is not set to UTF-8, it could cause problems.

Using a context manager is a good idea though.

**Comment by dbrgn** [comment]
> That said, I think we still want to work with 3.4 (for when Raspbian does ship it), but we need it to work for 3.2 (Wheezy stable)

Exactly, 3.2 should be the main goal right now, but eventually the library _will_ be used under Python 3.4.

I'd suggest we keep the versions 3.3 and 3.4 in tox, but remove them from Travis, what do you think? (Or we could flag them as allowed failures.)

File: setup.py (hunk 2 above)
Comment (doismellburning) [inline setup.py]: No :( (though I can't find the link I normally wave at people when this comes up - @ghickman got one handy?)

TL;DR `__init__` woe, filepath woe, general woe, sadness

Comment (dbrgn) [inline setup.py]: But now: happiness! 560ce6a734206bccb58343b248cf0216ccdfd178

Comment (ghickman) [inline setup.py]: Another option, if we want to have the version in the library (I'm not sure I see the point but definitely not against it), is to have a plain text `VERSION` the setup.py reads the version from then do the same in the codebase. It's still not ideal since we'll have some slight code duplication in our build and running code but unfortunately that's not something we can avoid with packaging right now.

**Comment by ghickman** [comment]
@dbrgn @doismellburning – No I think those are fair points, lets keep both versions in ready for when they're available.

**Comment by dbrgn** [comment]
OK, feedback has been integrated and the branch has been rebased.

Right now there's a build matrix on Travis with Python 2.7 and 3.2. There's also a build that does coverage testing, with an allowed failure. In there we could add a tool like Coveralls.

img

The tests [pass](https://travis-ci.org/py3minepi/py3minepi/jobs/36648632) on Python 2.7 and [fail](https://travis-ci.org/py3minepi/py3minepi/jobs/36648633) on all Python 3 versions.

In the "cov" build you can see an ASCII coverage report:

img

**Comment by doismellburning** [comment]
That Py3 failure looks...fun :s

**Comment by pozorvlak** [comment]
Yay for working CI builds! OTOH, it's a bit disappointing that it fails on
the Python dialect we're supposed to be targeting...

As for the QEMU thing: that was me following instructions from a random
blog post in a code sprint. We're currently doing a ridiculous number of
levels of virtualisation - QEMU within an ARM VM which is almost certainly
being simulated on x86 hardware - and as anyone who's seen _Inception_
knows, running a VM within a VM within a VM makes everything
veeeerrrrrryyyyy ssssslllllllooooooowwwwwww. If there's a faster way of
testing in our expected deployment environment, that would be awesome;
however, I'm reluctant to go a test setup that doesn't attempt to run at
least a few tests against the actual binary blob.

**Comment by doismellburning** [comment]
Yeah, it's slow and tedious but I definitely like the idea of Actual Realistic Testing - not sure I have any bright ideas about good ways to speed it though...

(DIY Jenkins testing against pre-built emulator images? That way at least we're pre-loading the work from most of https://github.com/py3minepi/py3minepi/blob/master/.travis-ci.sh, otoh that's a tedious maintenance requirement - I think I'd rather just have slower tests...)

**Comment by doismellburning** [comment]
Anyway this looks good. Can I confirm that this is a net improvement, i.e. some tests now pass whereas none did previously?

**Comment by pozorvlak** [comment]
+1 on it being a net improvement, and +1 for the overall idea of the change.

**Comment by ghickman** [comment]
I'm in agreement with @pozorvlak, it's a net improvement and adds some of the items we discussed in #14.

**Comment by dbrgn** [comment]
> That Py3 failure looks...fun :s

Probably some issue with identity comparison. But that test doesn't really matter anyways. Most of the code is still untested.

> If there's a faster way of testing in our expected deployment environment, that would be awesome

Yes, mocking. But mocking is always based on assumptions. We can just test what gets sent to Minecraft, not what Minecraft does with it.

> however, I'm reluctant to go a test setup that doesn't attempt to run at least a few tests against the actual binary blob.

While that sounds good, how do we actually write the assertions? There's no way to test whether a set_block command actually created a block. And Minecraft does not return any output on errors. So basically there's no way to test any assertions using the Minecraft binary.

**Comment by pozorvlak** [comment]
Why can't we send `getBlockWithData` calls to check if the block now has
the expected type?

**Comment by dbrgn** [comment]
Oh, true, we can do that... Sorry for forgetting :) It's an integration though, not a unit test. For unit tests, we should use mocking. It's faster anyways.

In that case I'll merge this.

**Comment by doismellburning** [comment]
Please never again merge your own pull request, thanks

**Comment by dbrgn** [comment]
@doismellburning I thought the three confirming comments above were a "good to merge". But in that case I'll leave it to others in the future :)

**Comment by pozorvlak** [comment]
@dbrgn: absolutely, anything that touches the Minecraft binary is an
integration test. But we should do integration tests in addition to our
unit tests, and the place to do so is in CI :-)
